modules/OFConnectionManager2/module/src/cxn_instance.h

```diff
@@ -0,0 +1,337 @@
+     * Once a message is read in, it is processed and the read buffer
+     * is cleared.
+     */
+    uint8_t read_buffer[READ_BUFFER_SIZE];
+    int read_bytes; /* Number of bytes currently in read buffer */
+    int bytes_needed; /* Num bytes needed for next process step */
+
+    /* Write queue */
+    bigring_t *write_queue; /* Ringbuffer of pointers to message data */
+    int write_queue_head_offset; /* Bytes already sent out from head of write_queue */
+    int bytes_enqueued;     /* Total bytes queued */
+    int pkts_enqueued;      /* Total pkts queued */
+
+    /* Additional debug info */
+    int has_debug_counters;
+    debug_counter_t tx_drop_counter;
+    debug_counter_t rx_counters[OF_MESSAGE_OBJECT_COUNT];
+    debug_counter_t tx_counters[OF_MESSAGE_OBJECT_COUNT];
+
+    /* Barrier */
+    int outstanding_op_cnt; /* Number of outstanding operations */
+    struct {
+        unsigned char pendingf;           /* Barrier reply pending flag */
+        uint32_t      xid;                /* XID of barrier request */
+    } barrier;
+
+    /* Echo */
+    struct {
+        uint32_t outstanding_echo_cnt;  /* number of unacknowledged echoes */
+        uint32_t threshold;  /* value above which cxn is declared dead */
+        uint32_t period_ms;  /* keepalive period in milliseconds */
+    } keepalive;
+
+    bundle_t bundles[MAX_BUNDLES];
+
+    /* Message Tracing */
+    aim_pvs_t* trace_pvs;
+
+    /* Used by the bsn_time_request message handler */
+    indigo_time_t hello_time;
+
+    /* Zero for main connection and non-zero for other connections */
+    uint8_t aux_id;
+
+    /* Pointer to the controller to which this connection belongs */
+    controller_t *controller;
+
+    /* If set, don't read any messages for this connection */
+    bool paused;
+
+    char desc[MAX_CONTROLLER_DESC_LEN+MAX_AUX_ID_DESC_LEN];  /* For logging */
+} connection_t;
+
+/**
+ * How many bytes in buffer are free
+ * See notes above about WRITE_BUFFER_SIZE.
+ */
+#define CXN_WRITE_BYTES_AVAIL(cxn) \
+    (WRITE_BUFFER_SIZE - ((cxn)->bytes_enqueued))
+
+/**
+ * Is a connection active (in any state)?
+ */
+#define CXN_ACTIVE(cxn) ((cxn)->active)
+
+/**
+ * Is connection marked local
+ */
+#define CXN_LOCAL(cxn) ((cxn)->controller->config_params.local)
+
+/**
+ * Is connection marked listen
+ */
+#define CXN_LISTEN(cxn) ((cxn)->controller->config_params.listen)
+
+/**
+ * Is a controller active?
+ */
+#define CONTROLLER_ACTIVE(controller) ((controller)->active)
+
+/**
+ * Is a connection TCP connected?
+ */
+#define CXN_TCP_CONNECTED(cxn)                                          \
+    (CXN_ACTIVE(cxn) &&                                                 \
+     ((cxn->state == CXN_S_HANDSHAKING) ||                              \
+      (cxn->state == CXN_S_HANDSHAKE_COMPLETE)))
+
+
+static inline indigo_cxn_config_params_t*
+get_connection_config(connection_t *cxn)
+{
+    if (cxn && cxn->controller) {
```

Comment:
When can this occur? Should this just be an assert so callers don't have to check the return value? Same for `get_connection_params` below.